tools/vectorlint/main.go

```diff
@@ -0,0 +1,190 @@
+// vectorlint analyzes vector files to flag potential issues
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"github.com/santhosh-tekuri/jsonschema/v6"
+	"io/fs"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+func main() {
+	schemaDirectory := flag.String("schemas-dir", "schemas", "directory containing schema files")
+	vectorsDirectories := flag.String("vectors-dir", "testvectors_v1,testvectors", "comma separated directories containing vector files")
+	vectorFilter := flag.String("vector-filter", "", "only validate vector files matching the provided pattern")
+
+	flag.Parse()
+
+	vectorDirectoryParts := strings.Split(*vectorsDirectories, ",")
+
+	log.Printf("reading schemas from %q\n", *schemaDirectory)
+	log.Printf("reading vectors from %q\n", vectorDirectoryParts)
+
+	var vectorRegex *regexp.Regexp
+	if *vectorFilter != "" {
+		vectorRegex = regexp.MustCompile(*vectorFilter)
+		log.Printf("filtering vectors with %q\n", *vectorFilter)
+	}
+
+	schemaCompiler := jsonschema.NewCompiler()
+
+	for _, f := range customFormats {
+		schemaCompiler.RegisterFormat(&f)
+	}
+
+	var total, valid, invalid, noSchema, ignored int
+	for _, vectorDir := range vectorDirectoryParts {
+		err := filepath.WalkDir(vectorDir, func(path string, d fs.DirEntry, err error) error {
+			if err != nil {
+				return err
+			}
+
+			if d.IsDir() || !strings.HasSuffix(d.Name(), ".json") {
+				return nil
+			}
+
+			if vectorRegex != nil && !vectorRegex.MatchString(d.Name()) {
+				return nil
+			}
+
+			vectorData, err := os.ReadFile(path)
+			if err != nil {
+				return fmt.Errorf("failed to read %s: %w", path, err)
+			}
+
+			total++
+
+			var vector struct {
+				Schema string `json:"schema"`
+			}
+
+			if err := json.Unmarshal(vectorData, &vector); err != nil {
+				log.Printf("❌ %q: invalid vector JSON data: %s\n", path, err)
+				invalid++
+				return nil
+			}
+
+			if vector.Schema == "" {
+				log.Printf("❌ %q: no schema specified\n", path)
+				noSchema++
+				return nil
+			}
+
+			if missingSchemas[vector.Schema] {
+				log.Printf("⚠️ %q: ignoring missing schema %q\n", path, vector.Schema)
+				ignored++
+				return nil
+			}
+
+			schemaPath := filepath.Join(*schemaDirectory, vector.Schema)
+			if _, err := os.Stat(schemaPath); os.IsNotExist(err) {
+				log.Printf("❌ %q: referenced schema %q not found\n", path, vector.Schema)
+				invalid++
+				return nil
+			}
+
+			schema, err := schemaCompiler.Compile(schemaPath)
+			if err != nil {
+				log.Printf("❌ %q: invalid schema %q: %s\n", path, vector.Schema, err)
+				invalid++
+				return nil
+			}
+
+			var instance any
+			_ = json.Unmarshal(vectorData, &instance)
```

Comment:
I'd still check this error. The previous Unmarshal should probably report all errors, but we might remove it or who knows what later.

Reply:
Done.